### PR TITLE
Drone

### PR DESCRIPTION
**Pull request by xoviat**

### CODE REVIEW

**Comment by michalpasztamobica** [comment]
Thanks for letting us know, @kiranchandramohan !

Hi @xoviat , I assume you want to use drone to get access to the arm architectures, which Github Actions sadly do not offer? Are you also targeting windows later on? Windows is available with Github Actions.
 
I think it would be good to stay with a single CI in the long run to avoid maintaining two scripts. Looking at this PR, they are similar when it comes to usage. Do you know if Drone can offer better performance (Github offers 2 cores for computation)? Maybe we should consider migrating existing scripts to Drone's machines instead of using Github Actions servers?

**Comment by xoviat** [comment]
The only reason to use drone is that it offers arm64 for free with no additional configuration (no setting up aws, etc.). The alternative is to run a github hosted instance on aws, but that would take far more of maintainer's time than this.

File: .github/workflows/drone.sh [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+flang --version
+CMAKE_OPTIONS="-DLLVM_TARGETS_TO_BUILD=OpenMP -DCMAKE_INSTALL_PREFIX=/usr/local -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_COMPILER=/usr/bin/gcc -DCMAKE_CXX_COMPILER=/usr/bin/g++"
+git clone --depth 1 --single-branch --branch release_90 https://github.com/llvm-mirror/openmp.git
+cd openmp
+mkdir -p build && cd build
+cmake $CMAKE_OPTIONS ..
+make -j$(nproc)
+make install
+mkdir -p build && cd build
+cmake -DLLVM_TARGETS_TO_BUILD=flang -DCMAKE_INSTALL_PREFIX=/usr/local -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_COMPILER=/usr/bin/gcc -DCMAKE_CXX_COMPILER=/usr/bin/g++ -DCMAKE_Fortran_COMPILER=/usr/local/bin/flang -DCMAKE_Fortran_COMPILER_ID=Flang -DFLANG_INCLUDE_DOCS=ON -DFLANG_LLVM_EXTENSIONS=ON ke -j$(nproc)
```
Comment (janisozaur) [inline .github/workflows/drone.sh:53]:
```suggestion
cmake -DLLVM_TARGETS_TO_BUILD=flang -DCMAKE_INSTALL_PREFIX=/usr/local -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_COMPILER=/usr/bin/gcc -DCMAKE_CXX_COMPILER=/usr/bin/g++ -DCMAKE_Fortran_COMPILER=/usr/local/bin/flang -DCMAKE_Fortran_COMPILER_ID=Flang -DFLANG_INCLUDE_DOCS=ON -DFLANG_LLVM_EXTENSIONS=ON ..
make -j$(nproc)
```

**Comment by xoviat** [comment]
@janisozaur Do you know what this means?

```
.github/workflows/drone.sh: 25: [: 3: unexpected operator
990
```

File: .github/workflows/drone.sh [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    echo "No artifacts in build $i, counting from latest" 
+    i=$[$i+1] 
+    wget --output-document artifacts_llvm `jq -r --argjson i "$i" '.workflow_runs[$i].artifacts_url?' runs_llvm.json` 
+    echo "cat artifacts_llvm" 
+    cat artifacts_llvm 
+done
+echo "cat artifacts_flang-driver"
+cat artifacts_flang-driver
+i="0"
+while [ `jq -r '.total_count?' artifacts_flang-driver` == "0" ] && [ $i -lt 3 ]
```
Comment (janisozaur) [inline .github/workflows/drone.sh:25]: @xoviat it means there's an error somewhere in this line. I expect something either evaluates to an empty string or something that's expected to be numeric isn't.

I suggest you add `set -x` before this line to show what it evaluates to

**Comment by kiranchandramohan** [comment]
@xoviat  Are you planning to continue working on this?

**Comment by xoviat** [comment]
@kiranchandramohan I cannot proceed further because `${{ secrets.GITHUB_TOKEN }}` does not exist in drone.

**Comment by kiranchandramohan** [comment]
@xoviat is GITHUB_TOKEN something I can help with?

**Comment by xoviat** [comment]
Yes. The GitHub script uses `GITHUB_TOKEN` to download the flang driver. Drone can't download the flang driver because that token is not available.

**Comment by michalpasztamobica** [comment]
Hi @xoviat , please take a look at [how we are using the GITHUB_TOKEN](https://github.com/flang-compiler/flang/blob/2168b3189ee3e04822d80b0fbd14dc3c0c062e2b/.github/workflows/build_flang.yml#L114).

Perhaps you need to add this to your `Authorisation Bearer`?

Another thing is how the `GITHUB_TOKEN` works... It should be available in the `yml` script as it is available for the existing Github Actions. 
If it is not available in drone's `sh` script, then maybe try passing it as an argument to `drone.sh` call?

**Comment by xoviat** [comment]
@michalpasztamobica GITHUB_TOKEN is not present in the environment: https://cloud.drone.io/flang-compiler/flang/22/1/2

Not sure what the solution is. See [this article](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token) for details on the GITHUB_TOKEN. It might need to be made a secret.

**Comment by kiranchandramohan** [comment]
@xoviat Is the issue specifically for the flang-driver repo? Or is it applicable to llvm also? BTW we are going to stop supporting llvm-9, so could you consider using classic-flang-llvm-project (release_11x)?

**Comment by xoviat** [comment]
The CI in this repository uses the GITHUB_TOKEN secret to download the driver. That secret is not present in the Drone CI. Therefore, the Drone CI cannot download the flang driver. Without the driver, flang cannot be compiled. Does that make sense?

**Comment by kiranchandramohan** [comment]
Is the same issue present for the llvm repo also? How did it work for llvm? Why is this present only for driver?

**Comment by xoviat** [comment]
Do you mean f18? It's only present for this repository because only this repository has to download the driver. The other repositories have all of the code they need (I think).

**Comment by xoviat** [comment]
More specifically, no other repository has this line:

```
wget --output-document flang-driver_build.zip --header="Authorization: Bearer " $url    
```

**Comment by kiranchandramohan** [comment]
Are you talking about the driver in https://github.com/flang-compiler/flang-driver ?
If so, you would also need https://github.com/flang-compiler/llvm 

My question is why is this showing up only for the flang-driver repo? Why isn't the line you are mentioning not present for the llvm repo? Also why is this a zip file?

**Comment by xoviat** [comment]
Neither of those repositories have CI enabled. So to qualify my statement, the other repositories **with CI** have all of the code they need (I think).

> Also, why is this a zip file?

Don't know. I copied it from here:

https://github.com/flang-compiler/flang/blame/master/.github/workflows/build_flang.yml#L100


Let me ask, do you see anything at https://github.com/flang-compiler/flang/settings/secrets/actions ?

**Comment by kiranchandramohan** [comment]
There are no secrets for this repository's environments.
There are no secrets for this repository.
No organization secrets have been authorized for this repository.

^ No secrets of any kind here. :)

**Comment by bryanpkc** [comment]
I believe that a GitHub workflow uses a system-generated token to identify itself to GitHub. That's why we don't need to define any explicitly.

When an external agent tries to access GitHub APIs, including downloading artifacts, the agent must provide a GitHub OAuth token to identify itself as a GitHub user or organization who is authorized to access the APIs or download the artifacts in question. I generated an OAuth token for myself at https://github.com/settings/tokens, and was then able to run the REST queries and download the ZIP file.

According to Drone documentation, when Drone is registered as a GitHub OAuth application in a user profile or in the organization, it will provide a "secret" which is in fact the OAuth token discussed above.

References:
- https://docs.drone.io/server/provider/github/
- https://docs.github.com/en/developers/apps/creating-an-oauth-app

**Comment by michalpasztamobica** [comment]
I managed to [download the artifacts](https://cloud.drone.io/michalpasztamobica/flang/75/1/2) form my fork of classic-flang-llvm-project with drone. It wasn't trivial for me, I has to ask for help on Drone's discourse. I believe we need to create an access token for flang-compiler/flang.

@xoviat , my solution (on my account: michalpasztamobica) is:

```
kind: pipeline
name: arm64_gcc_make

platform:
  os: linux
  arch: arm64

clone:
  disabled: true

steps:
- name: Build and Test
  image: ubuntu
  environment:
    GITHUB_TOKEN:
      from_secret: DRONE_GITHUB_TOKEN
  commands:
    - apt update
    - apt install -y wget
    - |
      wget --output-document output.file --header="Authorization: Bearer $${GITHUB_TOKEN}" https://api.github.com/repos/michalpasztamobica/classic-flang-llvm-project/actions/artifacts/47743037/zip
```

I tried with a final url, but this needs to be surrounded with the logic that finds this url.
This solution also required me to:
1) Create an access token in github, as @bryanpkc explained above.
2) Add the token (named `DRONE_GITHUB_TOKEN` in the example above) in drone settings, as @xoviat  [explained above](https://github.com/flang-compiler/flang/pull/993#issuecomment-803665342).

I did this in my account, so we need someone with admin access to `flang-compiler` to accomplish the same. @kiranchandramohan, @bryanpkc  would any of you be able to help?

@bryanpkc , perhaps setting drone as an OAuth app will also let us accomplish the task. If you find it a better solution for any reasons, I can try it out on my account. However I must say I find the solution described above equally good and proven to work.

**Comment by michalpasztamobica** [comment]
> Let me ask, do you see anything at https://github.com/flang-compiler/flang/settings/secrets/actions ?

In fact I think we need to add the secret to https://cloud.drone.io/flang-compiler/flang/settings/secrets

**Comment by xoviat** [comment]
I attempted to use a personal token to fetch the artifact but failed. Maybe it was done incorrectly, or maybe you have to have a certain level of access for your personal token to work. May I ask what stops users from printing the secret in a PR?

Also, I apologize for the amount of resources this PR has taken. If I had known it would be so much trouble, I probably wouldn't have bothered in the first place.

**Comment by michalpasztamobica** [comment]
First of all, there is no need to apologize. What you're working on is a valuable addition to the repo/project and we're happy to provide support. I must say configuring drone was not trivial to me either.

Answering your question. What I did is I used [my own account](https://github.com/michalpasztamobica/) for tests, not the `flang-compiler` organization one. I have my own fork of `classic-flang-llvm-project` and my own for of `flang`. In my account I have full rights to generate a token and add it to a repo. Token generated in my account will work for all repos in my account. It will not work on repos from different account.

It seems to me that you were trying to use your own `xoviat` personal token on the `flang-compiler`'s repo. 
As @bryanpkc  pointed out, for Github Actions there is no need to explicitly create a token. If you go `${{ secrets.GITHUB_TOKEN }}` the script will just get the token automatically - still it will only work within the same organization. Hence it is possible to authenticate a download of `flang-compiler/classic-flang-llvm-projects` artifacts in `flang-compiler/flang` script.

For drone I created a token in `michalpasztamobica` user/organization account and used it on `cloud.drone.io/michalpasztamobica` to authenticate a script from `michalpasztamobica/flang` to download artifacts from `michalpasztamobica/classic-flang-llvm-project`.

We now need to do the same, just for `flang-compiler` rather than `michalpasztamobica` :)

**Comment by kiranchandramohan** [comment]
I don't have access to the whole of flang-compiler organization. I think only @sscalpone or @gklimowicz has access.

I set up Drone as an OAuth application for my account and tried transferring the ownership to flang-compiler organization but it is pending.

**Comment by kiranchandramohan** [comment]
@xoviat can you try again?

**Comment by michalpasztamobica** [comment]
@bryanpkc , I didn't manage to run `wget` successfully in my forks after setting drone as an authenticated app in github. I set the app's address to `cloud.drone.io`, where it seems the jobs are running. I failed to download the artifact no matter if I used clientID, clientID+password or nothing at all for `wget` (I had to set user and password as secrets)
No idea what I am doing wrong this time, I can request support from Drone again. @bryanpkc , any ideas before I do that? Did you manage to set it up successfully on your account, perhaps?

@xoviat , is there a reason you closed this PR? Did you move the development somewhere else?

**Comment by bryanpkc** [comment]
@michalpasztamobica Sorry, I didn't set up any Drone account; I simply tried the GitHub OAuth token with wget/curl.

**Comment by michalpasztamobica** [comment]
Ok, no problem, perhaps I misunderstood something - I tried the same thing and it worked. It requires someone with Admin rights to generate a token on github and paste it as a secret to Drone. Is this an acceptable solution in the view of just authenticating the app not being sufficient or requiring extra investigation? @bryanpkc , @kiranchandramohan ?

**Comment by kiranchandramohan** [comment]
@michalpasztamobica is that for a particular repo or for the whole organization? I can do it for this repo and classic-flang-llvm-project but not for the whole organization.

**Comment by xoviat** [comment]
One alternative is to download the artifacts and post them publicly in a release. Then, the token is not required.

I closed this because it's become a distraction from the windows functionality. I'd like to focus on #980 if possible.

**Pull request closed without merging**